### PR TITLE
(POC): Dracut module for configuring HNV in the initrd

### PR DESCRIPTION
**Pull request by teclator**
This is just a revamp of https://github.com/agama-project/agama/pull/3221

- https://jira.suse.com/browse/PED-14534

# Add Hybrid Cloud Network (HCN) Support to Agama Live ISO

## Summary

This PR implements a new dracut module (`99hcnmgr`) that provides automatic Hybrid Cloud Network (HCN) configuration during the initramfs boot phase on IBM Power systems. This enables seamless network connectivity during live partition migration events on PowerPC (ppc64/ppc64le) architectures.

## Problem Statement

IBM Power systems use HCN (Hybrid Cloud Network) with bonded network interfaces to maintain connectivity during live partition migrations. When a partition migrates between hypervisors, the underlying network adapter changes (e.g., from SR-IOV to VNIC), but the bond interface ensures continuity. However, existing initramfs network configuration tools don't understand this PowerPC-specific device topology, requiring manual intervention.

## Solution

This PR introduces a dracut module that:

1. **Discovers HCN devices** from PowerPC device-tree (`/proc/device-tree`)
2. **Generates bonding configuration** automatically based on discovered topology
3. **Integrates with NetworkManager** via `nm-initrd-generator`
4. **Provides HCN-specific kernel parameters** (`rd.hcn.ip=`, `rd.hcn.route=`) for network configuration
5. **Persists network configuration** to the installed system
6. **Waits for device availability** after migration events (up to 3 minutes)
7. **Supports DNS configuration** via extended `rd.hcn.ip=` syntax

### Architecture

**Dracut Module Components:**

- **`module-setup.sh`**: Module registration and dependency declaration
  - PowerPC-only activation (`ppc64`/`ppc64le`) using standard `DRACUT_ARCH` with a fallback to `uname -m`
  - Clean systemd integration before NetworkManager starts
  - Declares explicit dependency on `"network-manager"` rather than a generic `"network"`
  - Only installs necessary binaries: `hcnmgr`, `hexdump`, `ofpathname`, `pseries_platform`, `awk`, `sed`
  - Bypasses the standard NetworkManager connection generator in built initramfs via post-install `sed` patches on NM services and hooks

- **`parse-hcnmgr.sh`**: Core discovery and configuration logic
  - Activates when `rd.hcn=1`, `rd.hcn.ip=`, or `rd.hcn.route=` is present (exits early if none found)
  - Scans device-tree for HCN-capable devices (PCI SR-IOV and VNICs)
  - Reads HCN ID, mode (primary/backup), and MAC addresses
  - Generates bond configuration with proper parameters (`mode=1,miimon=100,fail_over_mac=2`)
  - Processes HCN-specific `rd.hcn.ip=` and `rd.hcn.route=` parameters
  - Replaces slave device references with bond names automatically
  - Calls `nm-initrd-generator` to create NetworkManager connections
  - Fixes up generated connections to use bond names instead of UUIDs (required for `hcnmgr` compatibility)
  - Writes connections to `/etc/NetworkManager/system-connections/` for persistence
  - Adds NetworkManager finish hook to ensure the initqueue finishes successfully

- **`hcnmgr-initrd.service`**: Systemd unit for service orchestration
  - Runs before `NetworkManager-initrd.service` to ensure bonds exist before NetworkManager starts
  - Runs after `systemd-udev-trigger.service` to ensure devices are discovered
  - Conditional on `/proc/device-tree` existence (PowerPC-specific)
  - Includes `ConditionPathExists=/etc/initrd-release` to ensure it only evaluates in the initrd environment

### Key Features

**1. Device Discovery**

Scans PowerPC device-tree for HCN markers:

```bash
# Example device paths:
/proc/device-tree/pci@800000029008007/ethernet@0/ibm,hcn-id
/proc/device-tree/vdevice/vnic@30000006/ibm,hcn-id
```

Extracts:

- 4-byte HCN ID (used as bond name: `bond<ID>`)
- Mode (`primary` or `backup`) to set bond primary interface
- MAC address via `local-mac-address` property
- Interface name via `ofpathname` translation

**2. HCN Enabling and Configuration**

HCN can be enabled and configured using the following kernel parameters:

- **`rd.hcn=1`**: Enable HCN bond creation
  - Creates bond interfaces based on device-tree discovery
  - Bonds are created without IP configuration (configure separately if needed)

- **`rd.hcn.ip=`**: Enable HCN with IP configuration

  **DHCP on HCN bond:**

  ```bash
  rd.hcn.ip=dhcp
  # Generates: bond=bond333e80f5:...; ip=bond333e80f5:dhcp
  ```

  **DHCP with specific interface:**

  ```bash
  rd.hcn.ip=enP32775p1s0:dhcp
  # Discovers bond for enP32775p1s0, generates: ip=bond333e80f5:dhcp
  ```

  **Static IP configuration:**

  ```bash
  rd.hcn.ip=10.2.2.69::10.2.0.1:255.255.255.0
  # Generates: ip=10.2.2.69::10.2.0.1:255.255.255.0:bond333e80f5:none
  ```

  **Static IP with DNS:**

  ```bash
  rd.hcn.ip=10.2.2.69::10.2.0.1:255.255.255.0::none:8.8.8.8:8.8.4.4
  # Generates: ip=10.2.2.69::10.2.0.1:255.255.255.0:bond333e80f5:none:8.8.8.8:8.8.4.4
  ```

- **`rd.hcn.route=`**: Enable HCN with route configuration

  ```bash
  rd.hcn.route=192.168.1.0/24:10.2.0.1
  # Generates: rd.route=192.168.1.0/24:10.2.0.1:bond333e80f5
  ```

  **Route with specific interface:**

  ```bash
  rd.hcn.route=192.168.1.0/24:10.2.0.1:enP32775p1s0
  # Discovers bond for enP32775p1s0, generates: rd.route=192.168.1.0/24:10.2.0.1:bond333e80f5
  ```

**3. Persistent Network Configuration**

The module generates connections in `/etc/NetworkManager/system-connections/` (not just `/run`) to ensure they persist across boots. The Agama `save-agama-conf.sh` script has been enhanced to:

- Copy runtime connections generated by `nm-initrd-generator` (from `/run`)
- Copy persistent HCN connections (from `/etc`) to the installed system
- Mark HCN-configured networks as custom networks (via `/run/agama/custom_dracut_network`)

**4. NetworkManager Integration**

After calling `nm-initrd-generator`, the module performs critical connection fixups:

- Converts UUID-based `controller=` references to bond names (required for `hcnmgr` tool compatibility)
- Renames connection files from `<interface>.nmconnection` to `<bond>-<interface>.nmconnection`
- Ensures `port-type=bond` / `slave-type=bond` is set correctly
- Maps connections to bonds via interface name or MAC address matching

### New Files

- `live/live-root/usr/lib/dracut/modules.d/99hcnmgr/module-setup.sh` (49 lines)
- `live/live-root/usr/lib/dracut/modules.d/99hcnmgr/parse-hcnmgr.sh` (527 lines)
- `live/live-root/usr/lib/dracut/modules.d/99hcnmgr/hcnmgr-initrd.service` (19 lines)

### Modified Files

- `live/src/config.sh`: Added `hcnmgr` to dracut modules list
- `live/live-root/usr/lib/dracut/modules.d/99agama-cmdline/agama-network.sh`: Detect HCN parameters as custom network
- `live/live-root/usr/lib/dracut/modules.d/99agama-cmdline/save-agama-conf.sh`: Copy persistent HCN connections to installed system
- `live/live-root/usr/lib/dracut/modules.d/99agama-dud/agama-dud-parser.sh`: Minor shell syntax consistency
- `live/live-root/usr/lib/dracut/modules.d/99live-self-update/live-self-update-parser.sh`: Detect HCN parameters as custom network

### Test Suite (49 Comprehensive Tests)

Comprehensive BATS test suite under `live/test/hcnmgr/`:

**Unit Tests (`test_parse-hcnmgr.bats` - 35 tests):**

- Helper functions: `xdump4()`, `get_mac()`, `parse_nm_connection()`
- Connection fixup logic (UUID→bond name transformation, renaming, MAC matching)
- Command-line generation (`bond=`, `ip=`, `rd.route=` arguments) using HCN-specific parameters
- Edge cases: multiple bonds, missing modes, MAC address matching
- Performance & Quality validation: awk optimization, POSIX compliance, and fixture integrity validation

**Integration Tests (`test_integration.bats` - 14 tests):**

- End-to-end workflows with realistic device-tree fixtures
- Multi-bond configurations
- DHCP and static IP scenarios using HCN-specific parameters
- Route configuration
- DNS configuration with static IPs
- Error condition handling and compatibility tests
- Persistence verification (connections written to `/etc`)

**Test Fixtures:**

- Mock device-tree structure (`proc/device-tree/`)
- Mock sysfs (`sys/class/net/`)
- Sample NetworkManager connections (before/after fixup)
- Two complete bond scenarios: `bond333e80f5` and `bond5558a5a3`

**Test Infrastructure:**

- `run-tests.sh`: Convenient runner with colored output
- `Makefile`: CI integration targets
- `README.md`: Comprehensive test documentation
- `QUICK_REFERENCE.md`: Quick reference for HCN parameters and test usage
- `TEST_SUMMARY.md`: Overview and test coverage statistics

### Bond Creation Only

```bash
# Kernel cmdline
rd.hcn=1

# Result: Bond interfaces created from device-tree (no IP configuration)
```

### DHCP Configuration

```bash
# Kernel cmdline
rd.hcn.ip=dhcp

# Result: Automatic bond creation + DHCP on bond interface
```

### DHCP with Specific Interface

```bash
# Kernel cmdline
rd.hcn.ip=enP32775p1s0:dhcp

# Result: Bond created for enP32775p1s0's HCN, DHCP configured
```

### Static IP Configuration

```bash
# Kernel cmdline
rd.hcn.ip=10.2.2.69::10.2.0.1:255.255.255.0

# Result: Bond with static IP
```

### Static IP with DNS

```bash
# Kernel cmdline
rd.hcn.ip=10.2.2.69::10.2.0.1:255.255.255.0::none:8.8.8.8:8.8.4.4

# Result: Bond with static IP and DNS servers configured
```

### Static IP with Route

```bash
# Kernel cmdline
rd.hcn.ip=10.2.2.69::10.2.0.1:255.255.255.0 rd.hcn.route=192.168.1.0/24:10.2.0.1

# Result: Bond with static IP + custom route
```

### Multiple Bonds

```bash
# Module handles multiple HCN devices independently
# Each unique HCN ID gets its own bond
```

## Testing

Tested on PowerPC hardware with:

- SR-IOV PCI ethernet devices
- VNIC virtual network adapters
- Live partition migration scenarios
- Multiple bond configurations
- Persistence across boots and system installation

Run tests:

```bash
cd live/test/hcnmgr
./run-tests.sh
```

## Compatibility

- **Platforms**: PowerPC (ppc64, ppc64le) only
- **Required binaries**: `hexdump`, `ofpathname`, `pseries_platform`, `hcnmgr`, `awk`, `sed`
- **NetworkManager**: Uses `nm-initrd-generator` for connection creation
- **Dracut**: Tested with dracut-ng

## References

- `parse-hcnmgr-improvements.md`: Detailed refactoring documentation
- `live/test/hcnmgr/README.md`: Test suite documentation
- `live/test/hcnmgr/QUICK_REFERENCE.md`: Quick reference for HCN parameters
- `live/test/hcnmgr/TEST_SUMMARY.md`: Test coverage and statistics
- IBM PowerPC HCN documentation

## Migration from Manual Configuration

Previously, users needed to manually specify complex bond parameters:

```bash
bond=bond333e80f5:enP32775p1s0,env6:mode=1,... ip=10.2.2.69::...::bond333e80f5:off
```

Now, with this module, the user only specifies:

```bash
rd.hcn.ip=10.2.2.69::10.2.0.1:255.255.255.0
```

The module discovers devices, creates bonds, and maps configuration automatically.

---

This PR enables Agama to provide a seamless installation experience on IBM Power systems with HCN networking, handling live partition migrations transparently without user intervention. The configuration persists from the live ISO through to the installed system.

### CODE REVIEW

File: live/live-root/usr/lib/dracut/modules.d/99hcnmgr/hcnmgr-initrd.service [hunk 1]
```diff
@@ -0,0 +1,18 @@
+[Unit]
+Description=Hybrid Cloud Network Manager (Initrd)
+DefaultDependencies=no
+Wants=nm-initrd.service systemd-udev-trigger.service
```
Comment (aafeijoo-suse) [inline live/live-root/usr/lib/dracut/modules.d/99hcnmgr/hcnmgr-initrd.service]: IMO, this `Wants=nm-initrd.service` is too strong, and will cause a failure with recent versions of NetworkManager that ship `NetworkManager-initrd.service`, so remove it.

`nm-initrd.service` or `NetworkManager-initrd.service` are enabled in the initrd, so simply ordering `hcnmgr-initrd.service` before them (`Before=` below) should be enough.

File: live/live-root/usr/lib/dracut/modules.d/99hcnmgr/hcnmgr-initrd.service [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+After=systemd-udev-trigger.service
+Before=dracut-initqueue.service nm-initrd.service
```
Comment (aafeijoo-suse) [inline live/live-root/usr/lib/dracut/modules.d/99hcnmgr/hcnmgr-initrd.service]: Append `NetworkManager-initrd.service` to the `Before=` list.

File: live/live-root/usr/lib/dracut/modules.d/99hcn/hcn-init-initrd.service [hunk 3]
```diff
@@ -0,0 +1,18 @@
+[Unit]
+Description=Hybrid Cloud Network Manager (Initrd)
+DefaultDependencies=no
+Wants=nm-initrd.service systemd-udev-trigger.service
+After=systemd-udev-trigger.service
+Before=dracut-initqueue.service nm-initrd.service
+ConditionPathExists=/proc/device-tree
```
Comment (aafeijoo-suse) [inline live/live-root/usr/lib/dracut/modules.d/99hcn/hcn-init-initrd.service:8]: Prepend `ConditionPathExists=/etc/initrd-release`

File: live/live-root/usr/lib/dracut/modules.d/99hcnmgr/module-setup.sh [hunk 4]
```diff
@@ -0,0 +1,36 @@
+#!/bin/bash
+
+# HCN (Hybrid Cloud Network) dracut module setup
+
+# called by dracut
+check() {
+  # Only support PowerPC architectures
+  local _arch
+  _arch=$(uname -m)
```
Comment (aafeijoo-suse) [inline live/live-root/usr/lib/dracut/modules.d/99hcnmgr/module-setup.sh]: Check if `DRACUT_ARCH` is set: `local _arch=${DRACUT_ARCH:-$(uname -m)}`

File: live/live-root/usr/lib/dracut/modules.d/99hcnmgr/module-setup.sh [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+  [ "$_arch" = "ppc64" ] || [ "$_arch" = "ppc64le" ] || return 1
+
+  require_binaries hexdump ofpathname pseries_platform || return 1
+  return 0
+}
+
+# called by dracut
+depends() {
+  # This module depends on network support
+  echo "network"
```
Comment (aafeijoo-suse) [inline live/live-root/usr/lib/dracut/modules.d/99hcnmgr/module-setup.sh]: I think that the correct dependency here would be `network-manager` instead of just `network`, because it does not work with other network handlers (like `network-legacy` or `systemd-networkd`), right?

File: live/live-root/usr/lib/dracut/modules.d/99hcnmgr/parse-hcnmgr.sh [hunk 6]
```diff
@@ -0,0 +1,295 @@
...
+            print m_bond, m_iface; exit;
+          }
+        }
+      }
+    ')
+
+    if [ -n "$_mapping_info" ]; then
+      _found_master=${_mapping_info% *}
+      _found_ifname=${_mapping_info#* }
+    elif echo " $BOND_NAMES " | grep -q " ${_master:-$_controller} "; then
```
Comment (aafeijoo-suse) [inline live/live-root/usr/lib/dracut/modules.d/99hcnmgr/parse-hcnmgr.sh]: There are multiple `echo ... | grep -q ...` in this script, try to use `strstr` instead (function provided by `dracut-lib`), it does not spawn a `grep`. E.g.:
```sh
elif strstr "$BOND_NAMES" "${_master:-$_controller}"; then
```
For regular expressions like `if echo ":$HCN_IP:" | grep -qE ":($_s|$_s_dash):"; then`, you can use 2 conditions separated by "||".

File: live/live-root/usr/lib/dracut/modules.d/99hcnmgr/parse-hcnmgr.sh [hunk 7]
```diff
@@ -0,0 +1,295 @@
...
+
+    # Add bond definition to NEW_ARGS
+    BOND_OPTS="mode=1,miimon=100,fail_over_mac=2${PRIMARY:+,primary=$PRIMARY}"
+    NEW_ARGS="$NEW_ARGS bond=$BONDNAME:$SLAVES:$BOND_OPTS"
+
+    # Process hcn.ip - replace slave names/MACs with bond name
+    if [ -n "$HCN_IP" ]; then
+      _matched=0
+      for _s in $SLAVE_NAMES $SLAVE_MACS; do
+        _s_dash=$(echo "$_s" | tr ':' '-')
```
Comment (aafeijoo-suse) [inline live/live-root/usr/lib/dracut/modules.d/99hcnmgr/parse-hcnmgr.sh]: There are multiple `_s_dash=$(echo "$_s" | tr ':' '-')` in this script, replace all of them with `str_replace`  (function provided by `dracut-lib`) instead, it does not spawn a `tr`. E.g.:
```sh
_s_dash=$(str_replace "$_s" ":" "-")
```

File: live/live-root/usr/lib/dracut/modules.d/99hcnmgr/parse-hcnmgr.sh [hunk 8]
```diff
@@ -0,0 +1,295 @@
...
+        MOD_CMDLINE=$(echo "$MOD_CMDLINE" | sed -E "s/([:=])($_s|$_s_dash)([: ]|$)/\1$BONDNAME\3/g")
+        CHANGED=1
+      fi
+    done
+  done
+
+  # Write new configuration and update NetworkManager
+  if [ $CHANGED -eq 1 ] || [ -n "$NEW_ARGS" ]; then
+    info "parse-hcnmgr: writing /etc/cmdline.d/99-hcnmgr.conf"
+    mkdir -p /etc/cmdline.d
```
Comment (aafeijoo-suse) [inline live/live-root/usr/lib/dracut/modules.d/99hcnmgr/parse-hcnmgr.sh]: Remove this mkdir, `/etc/cmdline.d` is always created by dracut at build.

**Review by aafeijoo-suse — COMMENTED** [review]
Some superficial comments.

**Comment by aafeijoo-suse** [comment]
In order to use `rd.hcn=1` (or just `rd.hcn`) to  and keep `ip=` as is, try the following:

1. In `99hcnmgr/module-setup.sh`, before the end of the `install()` function, add:

```sh
[[ -e "${initdir}/var/lib/dracut/hooks/cmdline/99-nm-config.sh" ]] \
    && sed -i 's/nm_generate_connections/:/g' "${initdir}/var/lib/dracut/hooks/cmdline/99-nm-config.sh"
[[ -e "${initdir}/usr/lib/systemd/system/NetworkManager-config-initrd.service" ]] \
    && sed -i 's/\/usr\/libexec\/nm-initrd-generator/:/g' "${initdir}/usr/lib/systemd/system/NetworkManager-config-initrd.service"
[[ -e "${initdir}/usr/lib/systemd/system/NetworkManager-config-initrd.service.d/NetworkManager-config-initrd-dracut.conf" ]] \
    && sed -i -E 's/\/usr\/lib(exec)?\/nm-initrd-generator -- \$\(getcmdline\)/:/g' "${initdir}/usr/lib/systemd/system/NetworkManager-config-initrd.service.d/NetworkManager-config-initrd-dracut.conf"
```

2. In `99hcnmgr/parse-hcnmgr.sh`, before `info "parse-hcnmgr: starting"`, add:

```sh
if ! getargbool 0 rd.hcn; then
    warn "parse-hcnmgr: HCN not enabled"
    return 0
fi
```

3. In `99hcnmgr/parse-hcnmgr.sh`, parse `ip=` and `rd.route=` instead of `hcn.ip=` and `hcn.route=`:

```sh
HCN_IP=$(getargs ip)
HCN_ROUTE=$(getargs rd.route)
```

4. In `99hcnmgr/parse-hcnmgr.sh`, after the line with `: >/run/NetworkManager/initrd/neednet`, add:

```sh
echo '[ -f /tmp/nm.done ]' > "$hookdir"/initqueue/finished/nm.sh
```

File: live/live-root/usr/lib/dracut/modules.d/99hcnmgr/parse-hcnmgr.sh [hunk 9]
```diff
@@ -0,0 +1,408 @@
...
+  local conn_dir="/run/NetworkManager/system-connections"
+  [ -d "$conn_dir" ] || return 0
+
+  local con id uuid ifname master controller mac uuid_map
+  local found_master found_ifname mapping_info new_id
+
+  # First pass: build UUID to ID mapping for resolution
+  for con in "$conn_dir"/*.nmconnection; do
+    [ -e "$con" ] || continue
+    read -r id uuid ifname master controller mac <<EOF
```
Comment (cfconrad) [inline live/live-root/usr/lib/dracut/modules.d/99hcnmgr/parse-hcnmgr.sh]: I think we are missing a guard here, that the nmconnection files contain all these information.

File: live/live-root/usr/lib/dracut/modules.d/99hcnmgr/parse-hcnmgr.sh [hunk 10]
```diff
@@ -0,0 +1,408 @@
...
+      found_ifname=$ifname
+    fi
+
+    # Only modify connections that are HCN-related
+    if [ -n "$found_master" ]; then
+      info "hcnmgr: fixing up connection $id (HCN-related)"
+
+      # Ensure controller/master and port-type/slave-type are correct
+      if grep -q "^[[:space:]]*controller=" "$con"; then
+        sed -i "s/^[[:space:]]*controller[[:space:]]*=.*/controller=$found_master/" "$con"
```
Comment (cfconrad) [inline live/live-root/usr/lib/dracut/modules.d/99hcnmgr/parse-hcnmgr.sh]: Lets introduce getter and setter functions which are then better to read. Something like:

```bash
gkeyfile_get() {
    # Usage: gkeyfile_get <file> <section> <key>
    awk -F'=' -v sec="$2" -v key="$3" '
        $0 ~ "^\\[" { in_sec = ($0 == "[" sec "]") }
        in_sec && $1 == key { gsub(/[ \t]+$/, "", $2); gsub(/^[ \t]+/, "", $2); print $2; exit }
    ' "$1"
}

gkeyfile_set() {
    # Usage: gkeyfile_set <file> <section> <key> <val>
    if [ -n "$(gkeyfile_get "$1" "$2" "$3")" ]; then
            sed -i "/^\[$2\]/,/^\[/ s/^\([[:space:]]*$3[[:space:]]*=[[:space:]]*\).*/\1$4/" "$1"
    else
             sed -i "/^\[$2\]/a $3=$4" "$1"
    fi
}
```


```suggestion
      gkeyfile_set "$con" connection controller "$found_master"
```

File: live/live-root/usr/lib/dracut/modules.d/99hcnmgr/parse-hcnmgr.sh [hunk 11]
```diff
@@ -0,0 +1,408 @@
...
+        sed -i "/^\[connection\]/a controller=$found_master" "$con"
+      fi
+
+      if grep -q "^[[:space:]]*port-type=" "$con"; then
+        sed -i "s/^[[:space:]]*port-type[[:space:]]*=.*/port-type=bond/" "$con"
+      elif grep -q "^[[:space:]]*slave-type=" "$con"; then
+        sed -i "s/^[[:space:]]*slave-type[[:space:]]*=.*/slave-type=bond/" "$con"
+      else
+        sed -i "/^\[connection\]/a port-type=bond" "$con"
+      fi
```
Comment (cfconrad) [inline live/live-root/usr/lib/dracut/modules.d/99hcnmgr/parse-hcnmgr.sh]:
```suggestion
      if [ -n "$(gkeyfile_get "$con" connection slave-type)" ]; then
        gkeyfile_set "$con" connection slave-type bond"
      else
        gkeyfile_set "$con" connection port-type bond"
      fi
```

**Comment by aafeijoo-suse** [comment]
To avoid the the workaround of always disabling `nm_generate_connections` regardless of the `rd.hcn` command line option:

>     1. In `99hcnmgr/module-setup.sh`, before the end of the `install()` function, add:
> 
> ```shell
> [[ -e "${initdir}/var/lib/dracut/hooks/cmdline/99-nm-config.sh" ]] \
>     && sed -i 's/nm_generate_connections/:/g' "${initdir}/var/lib/dracut/hooks/cmdline/99-nm-config.sh"

Better use this sed in 99-nm-config.sh:

```sh
    && sed -i 's/if getargbool 0 rd.debug/if getargbool 0 rd.hcn; then\n    warn "rd.hcn=1: skipping nm-config"\n    return 0\nfi\n\n&/' "${initdir}/var/lib/dracut/hooks/cmdline/99-nm-config.sh"
```

> [[ -e "${initdir}/usr/lib/systemd/system/NetworkManager-config-initrd.service" ]] \
>     && sed -i 's/\/usr\/libexec\/nm-initrd-generator/:/g' "${initdir}/usr/lib/systemd/system/NetworkManager-config-initrd.service"

And this sed in NetworkManager-config-initrd.service:

```sh
    && sed -i 's/\[Service\]/ConditionKernelCommandLine=\!rd.hcn\n\n&/' "${initdir}/usr/lib/systemd/system/NetworkManager-config-initrd.service"
```

> [[ -e "${initdir}/usr/lib/systemd/system/NetworkManager-config-initrd.service.d/NetworkManager-config-initrd-dracut.conf" ]] \
>     && sed -i -E 's/\/usr\/lib(exec)?\/nm-initrd-generator -- \$\(getcmdline\)/:/g' "${initdir}/usr/lib/systemd/system/NetworkManager-config-initrd.service.d/NetworkManager-config-initrd-dracut.conf"
> ```

The latest condition and sed for NetworkManager-config-initrd-dracut.conf can be removed.

File: live/live-root/usr/lib/dracut/modules.d/99agama-dud/agama-dud-parser.sh [hunk 12]
```diff
@@ -18,7 +18,7 @@ fetch_updates() {
     if grep -q -s -E -e "https?://" -e "ftps?://" -e "smbs?://" -e "scp://" /tmp/agamadud.info; then
       echo "rd.neednet=1" > /etc/cmdline.d/agama-generated.conf
       # use DHCP if there is no network configuration provided by user
-      if ! getarg "ip="; then
+      if ! getarg "ip=" >/dev/null && ! getarg "hcn.ip=" >/dev/null; then
```
Comment (aafeijoo-suse) [inline live/live-root/usr/lib/dracut/modules.d/99agama-dud/agama-dud-parser.sh]: `hcn.ip` => `rd.hcn.ip`

File: live/live-root/usr/lib/dracut/modules.d/99hcn/parse-hcn.sh [hunk 13]
```diff
@@ -0,0 +1,530 @@
...
+#
+# Usage:
+#   rd.hcn=1              - Create bond connections only (no IP configuration)
+#   rd.hcn.ip=<config>    - Create bonds with HCN-specific IP configuration
+#   rd.hcn.route=<config> - Create bonds with HCN-specific route configuration
+
+command -v getargs >/dev/null || . /lib/dracut-lib.sh
+
+# Check if HCN is enabled via any of the three methods
+if ! getargbool 0 rd.hcn && ! getargs rd.hcn.ip >/dev/null && ! getargs rd.hcn.route >/dev/null; then
```
Comment (aafeijoo-suse) [inline live/live-root/usr/lib/dracut/modules.d/99hcn/parse-hcn.sh]: This check is redundant, the call to this script is already filtered by `ConditionKernelCommandLine=` properties in `hcn-init-initrd.service` (unless there are cases where this script can be called directly and not through systemd).

File: live/live-root/usr/lib/dracut/modules.d/99hcn/parse-hcn.sh [hunk 14]
```diff
@@ -0,0 +1,530 @@
...
+  info "parse-hcn: HCN not enabled (no rd.hcn=1, rd.hcn.ip, or rd.hcn.route), skipping"
+  if (return 0 2>/dev/null); then
```
Comment (aafeijoo-suse) [inline live/live-root/usr/lib/dracut/modules.d/99hcn/parse-hcn.sh]: What does gemini want here?

File: live/live-root/usr/lib/dracut/modules.d/99hcn/parse-hcn.sh (hunk 13 above)
Comment (teclator) [reply live/live-root/usr/lib/dracut/modules.d/99hcn/parse-hcn.sh]: Exactly, It didn't hurt but it is already covered by the service

File: live/live-root/usr/lib/dracut/modules.d/99hcn/parse-hcn.sh (hunk 14 above)
Comment (teclator) [reply live/live-root/usr/lib/dracut/modules.d/99hcn/parse-hcn.sh]: It was done by Claude in this case ;). I guess it is a common pattern in dracut modules where the script can be sourced as library or executed as an standalone script. although you probably know better than me in that regard ;). We are not interested in the first although the logic for detecting the HCN devices could be moved to a library.. BTW, will remove the check completely.

**Review by ancorgs — APPROVED** [review]
Too much code and specialized knowledge for me to do a proper review.

But I see that, at least, @aafeijoo-suse and @teclator were involved. Based on that, I'm approving this.
